### PR TITLE
Use dask-spec for SSHCluster

### LINKED ISSUE #3903: SSHCluster can't turn dashboard off
Trying to turn off the dashboard results in a traceback at launch for SSHCluster
```python
from distributed import SSHCluster
cluster = SSHCluster(['localhost']*4, scheduler_options={'dashboard_address':None})
```
<details>
  <summary>Traceback - Click to expand!</summary>
  
```
distributed.deploy.ssh - INFO - distributed.scheduler - INFO - -----------------------------------------------
distributed.deploy.ssh - INFO - distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
distributed.deploy.ssh - INFO - Traceback (most recent call last):
distributed.deploy.ssh - INFO - File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/node.py", line 232, in start_http_server
distributed.deploy.ssh - INFO - self.http_server.listen(**http_address)
distributed.deploy.ssh - INFO - File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/tornado/tcpserver.py", line 151, in listen
distributed.deploy.ssh - INFO - sockets = bind_sockets(port, address=address)
distributed.deploy.ssh - INFO - File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/tornado/netutil.py", line 117, in bind_sockets
distributed.deploy.ssh - INFO - socket.getaddrinfo(address, port, family, socket.SOCK_STREAM, 0, flags),
distributed.deploy.ssh - INFO - File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/socket.py", line 918, in getaddrinfo
distributed.deploy.ssh - INFO - for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
distributed.deploy.ssh - INFO - socket.gaierror: [Errno 8] nodename nor servname provided, or not known
Traceback (most recent call last):
  File "/Users/shannon/Library/Application Support/JetBrains/PyCharm2020.1/scratches/scratch_1.py", line 19, in <module>
    cluster = SSHCluster(['localhost']*4, scheduler_options={'dashboard_address':None})
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/ssh.py", line 336, in SSHCluster
    return SpecCluster(workers, scheduler, name="SSHCluster", **kwargs)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/spec.py", line 256, in __init__
    self.sync(self._start)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/cluster.py", line 163, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/utils.py", line 347, in sync
    raise exc.with_traceback(tb)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/utils.py", line 331, in f
    result[0] = yield future
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/spec.py", line 284, in _start
    self.scheduler = await self.scheduler
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/spec.py", line 50, in _
    await self.start()
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/ssh.py", line 203, in start
    raise Exception("Worker failed to start")
Exception: Worker failed to start
distributed.deploy.spec - WARNING - Cluster closed without starting up
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/utils.py", line 198, in ignoring
    yield
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/spec.py", line 611, in close_clusters
    cluster.close(timeout=10)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/cluster.py", line 84, in close
    return self.sync(self._close, callback_timeout=timeout)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/cluster.py", line 163, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/utils.py", line 347, in sync
    raise exc.with_traceback(tb)
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/utils.py", line 331, in f
    result[0] = yield future
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/spec.py", line 400, in _close
    await super()._close()
  File "/Users/shannon/opt/miniconda3/envs/pywakes/lib/python3.8/site-packages/distributed/deploy/cluster.py", line 73, in _close
    await self._watch_worker_status_comm.close()
AttributeError: 'SpecCluster' object has no attribute '_watch_worker_status_comm'
```

</details>

**Environment**:

- Dask version: 2.17.2
- Python version: 3.8.3
- Operating System: macOS
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by charlesbluca**
Revisiting #3907; as said there:

> Previously we used to specify CLI commands like dask-worker. Now we specify Python classes like Worker or Nanny. This makes it more consistent to pass through keywords without having to go from Python to CLI back to Python.
> 
> However, this will also probably force dask-cuda to push its logic from a CLI command into a Python class. This commit speculatively changes documentation here in order to suggest a CUDAWorker class type.
> 
> I think that using dask-spec is probably a clean way forward generally. Most systems that have switched over to it tend to have fewer complaints.

This is essentially the same PR, but with an additional [suggestion ](https://github.com/dask/distributed/pull/3907#discussion_r445858270)from @pentschev to cast worker names to string in `SpecCluster._correct_state_internal()`. We will probably need to make minor changes in dask-cuda to handle the switch from `worker_module` -> `worker_class`

- [ ] Closes #3903, #3907
- [ ] Tests added / passed
- [ ] Passes `black distributed` / `flake8 distributed` / `isort distributed`

### CODE REVIEW

**Comment by pentschev** [comment]
Thanks @charlesbluca for reviving this. As for the Dask-CUDA changes, I have https://github.com/rapidsai/dask-cuda/pull/326/files open that was waiting for https://github.com/dask/distributed/pull/3907. If this is accepted, I'll update it to resolve that. 🙂

**Comment by charlesbluca** [comment]
It seems like the string casting fix *shouldn't* be needed anymore, as rapidsai/dask-cuda#656 began casting the name to string on dask-cuda's end. However, when this fix is excluded, starting an `SSHCluster` silently fails to spawn any workers:

```python
cluster = SSHCluster(  # makes it to dask_cuda.initialize preload
    ["localhost", "localhost"],
    connect_options={"known_hosts": None},
    worker_class="dask_cuda.CUDAWorker"
)
cluster.scheduler_info["workers"]  # {}
```

Not sure why this is. The string fix itself is the cause of the test failures, as it causes a mismatch in the worker names between `SpecCluster.workers` and `SpecCluster.scheduler_info["workers"]`

```python
cluster.workers  # {0: <SSH Worker: status=Status.running>}
{d["name"] for d in cluster.scheduler_info["workers"].values()}  # {'0'}  
```

Which seemingly prevents workers from being removed properly:

https://github.com/dask/distributed/blob/ac55f25d230c7144ed618d1d4374f254303a4e0a/distributed/deploy/spec.py#L368-L382

This can probably be solved by making a copy of `self.scheduler_info["workers"][msg]["name"]` that's cast as an int when accessing `self.workers`, but we would need a special case here for dask-cuda since its worker names are of the form `"0-N"`.

**Comment by charlesbluca** [comment]
> However, when this fix is excluded, starting an SSHCluster silently fails to spawn any workers:

After chatting with @pentschev, we now understand that this was because of a vague conditional on Dask-CUDA's end - that should be an easy fix.

Since Dask-CUDA is already casting the worker's name to string, I recommend we merge this PR as is (assuming all tests pass); if anything, we can make a follow up PR managing the worker names if it becomes a larger issue.

File: distributed/deploy/ssh.py
Comment (jacobtomlinson) [inline distributed/deploy/ssh.py:82]: Should this be a `DeprecationError`?

File: distributed/deploy/ssh.py
Comment (jacobtomlinson) [inline distributed/deploy/ssh.py:124]: In some other implementations we also serialize the local config and pass it along in an environment variable. Might be worth doing the same here.

https://github.com/dask/dask-cloudprovider/blob/fcc6cf37933d5098d3f0a9ed55e3437ca3cd6e98/dask_cloudprovider/generic/vmcluster.py#L33-L35

File: distributed/deploy/ssh.py
Comment (jacobtomlinson) [inline distributed/deploy/ssh.py:330]: Same here should maybe be a deprecation error.

File: distributed/deploy/ssh.py
Comment (charlesbluca) [reply distributed/deploy/ssh.py:82]: I could change this to a `DeprecationWarning`, though I'm not sure if it would confuse users to get a warning that interrupts the program. In the original PR, it seemed like @mrocklin [preferred ](https://github.com/dask/distributed/pull/3907#discussion_r440941318) raising an explicit error rather than a warning so that we wouldn't have to support both `worker_module` and `worker_class`.

From that discussion, it seems like the main motive behind `worker_module` was for dask-cuda benchmarks, so once this and rapidsai/dask-cuda#326 are merged we should be able to remove this error and the `worker_module` kwarg altogether (though we'll probably want to wait a bit to see if there are any downstream breakages from this).

File: distributed/deploy/ssh.py
Comment (charlesbluca) [reply distributed/deploy/ssh.py:124]: Looks like that might already be happening a few lines above?

https://github.com/dask/distributed/blob/31afb54cf02141050ae1cc7cbd7c3919557637fe/distributed/deploy/ssh.py#L93-L106

Comment (jacobtomlinson) [reply distributed/deploy/ssh.py:124]: Awesome! I totally missed that, didn't expand the diff properly.

File: distributed/deploy/ssh.py
Comment (jacobtomlinson) [reply distributed/deploy/ssh.py:82]: Fair enough.

**Review by jacobtomlinson — APPROVED** [review]
Sorry for the delay here. Thanks @charlesbluca.

**Comment by condoratberlin** [comment]
I just recognised this change after updating distributed.
Now my workflow is broken.
I use the SSHCluster to deploy a cluster on a HPC-Cluster. I allocate several nodes with 64cores each. My approach was to start a wokers with the nprocs command. So the length of the hosts list equals the number of nodes (+1 for the scheduler) and i started 64 processes there.
Now I had to enlarge the number of hosts to create the same number of workers, but for every worker a ssh connection has to be established to start them. Now the problem is, that the number of ssh connections is limited to ~50 and starting the high number of workers failed.
Are there any suggestions how start multiple workers with a single ssh connection?

**Comment by condoratberlin** [comment]
Sure, I create it on Monday.

**Pull request merged**
